src/context.rs

```diff
@@ -1,14 +1,20 @@
 use device::DeviceInfo;
 use ffi;
+use ffi::MaybeError;
 use io::{InputPort, OutputPort};
 use std::os::raw::c_int;
+use std::ptr;
+use std::ffi::CString;
+use std::sync::Mutex;
 use types::{Error, PortMidiDeviceId, Result};
 
 /// The PortMidi base struct.
 /// Initializes PortMidi on creation and terminates it on drop.
 pub struct PortMidi {
     device_count: u32,
+    virtual_devs: Mutex<Vec<PortMidiDeviceId>>
```

Comment:
I don't see why you use a Mutex? If it's to make the PortMidi Struct send and sync, the best is to let the user decide and wrap the context in a Mutex or anything he prefers.

Reply:
I use a mutex for making the `virtual_devices.rs` example work. 
If I get you right you suggest making the user wrap the PortMidi struct inside an `Arc<Mutex<>>` in a whole for using it inside different threads? 

Wouldn't that make access to the non-mutable methods unnecessarily complex? In every case, I have to make sure the `virtual_devs` member and the corresponding methods are not accessed by different threads at the same time as they are mutable.

Reply:
I said that because PortMidi struct can be used in a multi thread env, async or mono thread. It depends on the user needs, not the lib one. For example, why you use a Mutex and not a RwLock? It depends on how the lib is used. If you define a mutex everyone will have to manage this mutex even if there's no need. There's a cost at using a mutex.
If you think that locking the mutex is complex, define your own struct and hide the locking call in it.
Why you want interior mutability? create_virtual_device can use a self mutable, it's even more logic that you mutate thing when creating a device. Interior mutability should only be use when you want to hide an internal algo that need mutability on a function that shouldn't be mutable like a counter on a get.
In my opinion, the fact that you keep the id of the created virutal device is not part of the lib, but more how the developer use the lib. If I create only one device and don't need to keep it. You should remove the virtual_devs ids vector and only provide PortMidi function.
Last thought, what you should add to PortMidi lib is to call Pm_DeleteVirtualDevice when the return Id is dropped. It allows easing the management of the devices.